R/Algorithms.R

```diff
@@ -167,3 +167,74 @@ one_plus_lambda_EA <- function(IOHproblem, lambda_ = 1, budget = NULL) {
   }
   return(best_value)
 }
+
+
+#' One-Comma-Lambda Genetic Algorithm with 2-rate self-adaptive mutation rate
+#' 
+#' A genetic algorithm that controls the mutation rate (strength) using the so-called 
+#' 2-rate self-adaptation mechanism: the mutation rate is based on a parameter r. For
+#' each generation, half offspring are generated by mutation rate 2r/dim, and half by 
+#' r/2dim. r that the best offspring has been created with will be inherited by 
+#' probability 3/4, the other by 1/4.
+#'
+#' @param IOHproblem An IOHproblem object
+#' @param lambda_ integer, the population size > 1
+#' @param budget How many times the objective function can be evaluated
+#'
+#' @export
+#' @examples
+two_rate_GA <- function(IOHproblem, lambda_ = 2, budget = NULL){
+  dim = IOHproblem$dimension
+  if (is.null(budget)) budget <- 100*dim
+  obj <- IOHproblem$obj_func
+
+  parent <- sample(c(0, 1), dim, TRUE)
+  best <- parent
+  r <- 2.0
+  best_value <- obj(parent)
+  budget <- budget-1
+
+  while ( budget > 0 ){
+    selected_r <- r;
+    selected_obj <- Inf
+
+    for (i in 1:lambda_){
+      offspring <- parent
+
+      if(i <= lambda_/2){
+        mutation_rate = r / 2.0 / dim;
+      } else{
+        mutation_rate = 2.0 * r / dim;
+      }
+      offspring <- mutate(offspring, mutation_rate)
+
+      v <- obj(offspring)
+      if(v >= best_value){
+        best_value <- v
+        best <- offspring
+      }
+      if(v >= selected_obj){
+        selected_obj = v
+        selected_r = mutation_rate * dim;
+      }
+      budget <- budget - 1
+      if (budget == 0 ) break
+    }
+    parent <- best
+    if(runif(1) > 0.5){
+      r = selected_r
+    }
+    else{
+      if(runif(1) > 0.5){
+        r = r / 2.0
+      } else{
+        r = 2.0 * r
+      }
+
+      if(r < 2.0) r = 2.0
```

Comment:
Shouldnt this check be outside the else, otherwise you can end up with selected r of 1/2 * dimension?